File: python/find_the_difference_of_two_arrays.py
class Solution:
    def findDifference(self, nums1: list[int], nums2: list[int]) -> list[list[int]]:
        result: list[set] = [set(), set()]
        nums1Set: set[int] = set(nums1)
        nums2Set: set[int] = set(nums2)
        for i in range(len(nums1)):
            if nums1[i] not in nums2Set:
                result[0].add(nums1[i])
        for i in range(len(nums2)):
            if nums2[i] not in nums1Set:
                result[1].add(nums2[i])

        return [list(result[0]), list(result[1])]


def test_findDifference_case_1():
    # arrange
    nums1: list[int] = [1, 2, 3]
    nums2: list[int] = [2, 4, 6]
    expected: list[list[int]] = [[1, 3], [4, 6]]

    # act
    solution = Solution()
    actual = solution.findDifference(nums1, nums2)

    # assert
    assert expected == actual


def test_findDifference_case_2():
    # arrange
    nums1: list[int] = [1, 2, 3, 3]
    nums2: list[int] = [1, 1, 2, 2]
    expected: list[list[int]] = [[3], []]

    # act
    solution = Solution()
    actual = solution.findDifference(nums1, nums2)

    # assert
    assert expected == actual

File: rust/src/code/find_the_difference_of_two_arrays.rs
struct Solution {}

impl Solution {
    pub fn find_difference(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<Vec<i32>> {
        use std::collections::HashSet;

        let mut result: Vec<HashSet<i32>> = vec![HashSet::new(), HashSet::new()];
        let mut nums1_set: HashSet<i32> = HashSet::new();
        let mut nums2_set: HashSet<i32> = HashSet::new();

        for i in 0..nums1.len() {
            nums1_set.insert(nums1[i]);
        }
        for i in 0..nums2.len() {
            nums2_set.insert(nums2[i]);
        }

        for i in 0..nums1.len() {
            if nums2_set.contains(&nums1[i]) {
                continue;
            }
            result[0].insert(nums1[i]);
        }
        for i in 0..nums2.len() {
            if nums1_set.contains(&nums2[i]) {
                continue;
            }
            result[1].insert(nums2[i]);
        }

        vec![
            result[0].clone().into_iter().collect::<Vec<i32>>(),
            result[1].clone().into_iter().collect::<Vec<i32>>(),
        ]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_find_difference_case_1() {
        // arrange
        let nums1: Vec<i32> = vec![1, 2, 3];
        let nums2: Vec<i32> = vec![2, 4, 6];
        let mut expected: Vec<Vec<i32>> = vec![vec![1, 3], vec![4, 6]];

        // act
        let mut actual = Solution::find_difference(nums1, nums2);

        // assert
        expected.sort();
        actual.sort();
        expected.iter_mut().for_each(|x| x.sort());
        actual.iter_mut().for_each(|x| x.sort());
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_find_difference_case_2() {
        // arrange
        let nums1: Vec<i32> = vec![1, 2, 3, 3];
        let nums2: Vec<i32> = vec![1, 1, 2, 2];
        let mut expected: Vec<Vec<i32>> = vec![vec![3], vec![]];

        // act
        let mut actual = Solution::find_difference(nums1, nums2);

        // assert
        expected.sort();
        actual.sort();
        expected.iter_mut().for_each(|x| x.sort());
        actual.iter_mut().for_each(|x| x.sort());
        assert_eq!(expected, actual);
    }
}
